### ISSUE TITLE
Where should cops that might break things go?

### ISSUE THREAD

**Issue by DNNX** [bbatsov/rubocop#2606]
**TLDR**: there are cops that are very useful but can potentially break code. Where should we put them? `config/disabled.yml`? Separate gem (say, `violent_cops` or something)? Special category (say `Hardcore`) and then add a new flag `rubocop --hardcore` which is disabled by default?

**Long story**

I have a couple of ideas of new cops that might be useful (especially with autocorrection enabled), but the problem is that the autocorrection can break things in edge cases. Here are some examples:

The recent `str.start_with?(x) || str.start_with?(y)` => `str.start_with?(x, y)`. We _have to_ make sure that `y` is not a method call, otherwise the change is not equivalent to the original code. However, in most cases `y` is just a reader method which is safe to call any number of times. But we cannot know it for sure in Rubocop.

`1+rand(expr)` => `rand(1..expr)` transformation. [Happens](https://github.com/ManageIQ/manageiq/search?utf8=%E2%9C%93&q=rand) [all](https://github.com/solidusio/solidus/blob/0f06aa107b34401b32046c97ffc196f5fd3873eb/core/lib/spree/testing_support/sequences.rb#L5) [the](https://github.com/OpsTree/Puppet/blob/ea26d25bbb1111823e7e79098e6ae8e9dc04336e/modules/WordPressLEMPsetup/stdlib/lib/puppet/parser/functions/shuffle.rb#L35) [time](https://github.com/garywong89/phase-0/blob/a4a9fb6ceee9997ea60d4bd2858d47b7981fd06c/week-8/ruby/fizz_buzz_spec.rb#L7). The thing is, `expr` can be not an int but it can be a range, or a double, or even nil. We can safely apply this autocorrection only if it is an `Integer`.

Except we can't. If `expr's` value is `0`, then the source code returns a float within `1.0...2.0` range (which is unlikely the desired result), but the autocorrected `rand(1..0)` version returns `nil` (don't think it was the intention too). Negative ints have different behavior too.

Another, Rails-specific example is `x.try(y)`. In my experience, in almost all cases it should be replaced by `x.try!(y)` which has simpler semantics and works faster. Or even this shiny new `.?` thing. But of course we cannot autocorrect this because Rubocop cannot infer whether `x` can possibly be a non-nil object that does not respond to `y`.

Rubocop is a widely used gem, so we have to be conservative in order not to break other people's code. However, if I'm sure that have good tests, I want to turn on hardcore mode and correct as much things as possible automatically. Unlike other cops, the new hardcore cops are free assume that I'm a sane person and don't override core methods like "#to_i", that I don't have side effects in `start_with?` arguments and receiver, and so one.

**Comment by alexdowad** [comment]
Just to add a small point to the discussion: one option is to warn, but not auto-correct.

**Comment by Drenmi** [comment]
There are a few "dangerous" cops already, and it often boils down to what you mentioned, of not being able to know what a receiver will be at runtime. 🙂 So far the treatment has been on a case-by-case basis. You have a couple of options. You can:

 - disable the cop by default,
 - disable auto-correct by default,
 - make the `#autocorrect` method able to detect "unsafe" corrections,
 - decide not to give the cop auto-correct,
 - provide configuration options to reduce false positives.

These are all valid answers in different situations. You'll need to pick the one that makes most sense for your cop. 🙂 As a rule of thumb, false positives are unavoidable, and fine to have, because there are tools to squelch them, but there's a limit. If your cop generates hundreds of false positives in a standard codebase, then it's not usable. Note also that false negatives are much worse, because even if the user becomes aware of them, there's nothing they can do to make the cop smarter.

Right now cops are grouped roughly by domain (style, performance, third party library, etc.) I think it would be confusing to start grouping by hazardousness too.

Hope that answers your question. 😊

**Comment by Drenmi** [comment]
@jonas054, @bbatsov: If the question is sufficiently answered, I guess we can close this. 🙂
